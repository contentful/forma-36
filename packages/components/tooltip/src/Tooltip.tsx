import React, {
  useEffect,
  useState,
  useRef,
  MouseEvent,
  FocusEvent,
  CSSProperties,
} from 'react';
import { usePopper } from 'react-popper';
import { Placement } from '@popperjs/core';
import { cx } from 'emotion';
import type * as CSS from 'csstype';

import { styles } from './Tooltip.styles';

import tokens from '@contentful/f36-tokens';

import { Portal } from '@contentful/f36-utils';
import type { CommonProps } from '@contentful/f36-core';
import { Primitive } from '@contentful/f36-core';

export type TooltipPlace = Placement;

export interface TooltipProps extends CommonProps {
  /**
   * Child nodes to be rendered in the component and that will show the tooltip when they are hovered
   */
  children: React.ReactNode;
  /**
   * HTML element used to wrap the target of the Tooltip
   */
  as?: React.ElementType;
  /**
   * Content of the Tooltip
   */
  content?: React.ReactNode;
  id?: string;
  /**
   * It controls the initial visibility of the Tooltip
   */
  isVisible?: boolean;
  /**
   * It sets a max-width for the Tooltip
   */
  maxWidth?: number | CSS.Property.MaxWidth;
  /**
   * It sets a delay period for the Tooltip
   */
  hideDelay?: number;
  /**
   * Function that will be called when target gets blurred
   */
  onBlur?: (evt: FocusEvent) => void;
  /**
   * Function that will be called when target gets focused
   */
  onFocus?: (evt: FocusEvent) => void;
  /**
   * Function that will be called when the user move the mouse out of the target
   */
  onMouseLeave?: (evt: MouseEvent) => void;
  /**
   * Function that will be called when the user move the mouse over of the target
   */
  onMouseOver?: (evt: MouseEvent) => void;
  /**
   * Function that will be called when the user uses a keyboard key on the target
   */
  onKeyDown?: (evt: KeyboardEvent) => void;

  /**
   * It sets the "preferred" position of the Tooltip
   */
  placement?: TooltipPlace;
  /**
   * Class names to be appended to the className prop of the Tooltipâ€™s target
   */
  targetWrapperClassName?: string;
  /**
   * Boolean to control whether or not to render the tooltip in a React Portal.
   * Rendering content inside a Portal allows the tooltip to escape the bounds
   * of its parent while still being positioned correctly. Using a Portal is
   * necessary if an ancestor of the tooltip hides overflow.
   *
   * Defaults to `false`
   */
  usePortal?: boolean;
}

export const Tooltip = ({
  children,
  className,
  as: HtmlTag = 'span',
  content,
  id = 'cf-ui-tooltip',
  isVisible = false,
  hideDelay = 0,
  onBlur,
  onFocus,
  onMouseLeave,
  onMouseOver,
  onKeyDown,
  targetWrapperClassName,
  maxWidth = 360,
  testId = 'cf-ui-tooltip',
  placement = 'auto',
  usePortal = false,
  ...otherProps
}: TooltipProps) => {
  const [show, setShow] = useState(isVisible);

  const elementRef = useRef(null);
  const popperRef = useRef(null);
  const [arrowRef, setArrowRef] = useState<HTMLSpanElement | null>(null);
  const { styles: popperStyles, attributes, forceUpdate } = usePopper(
    elementRef.current,
    popperRef.current,
    {
      placement: placement,
      modifiers: [
        {
          name: 'arrow',
          options: {
            element: arrowRef,
            padding: parseFloat(tokens.borderRadiusSmall),
          },
        },
        {
          name: 'offset',
          options: {
            offset: [0, 10],
          },
        },
      ],
    },
  );

  // necessary to update tooltip position in case the content is being updated
  useEffect(() => {
    if (forceUpdate !== null) {
      forceUpdate();
    }
  }, [content, forceUpdate]);

  const [isHoveringTarget, setIsHoveringTarget] = useState(false);
  const [isHoveringContent, setIsHoveringContent] = useState(false);
  useEffect(() => {
    setShow(isHoveringContent || isHoveringTarget);
  }, [isHoveringTarget, isHoveringContent]);

  const contentMaxWidth =
    typeof maxWidth === 'string' ? maxWidth : `${maxWidth}px`;

  const contentStyles: CSSProperties = {
    zIndex: tokens.zIndexTooltip,
    maxWidth: contentMaxWidth,
    ...popperStyles.popper,
  };

  if (!content) {
    return <HtmlTag className={targetWrapperClassName}>{children}</HtmlTag>;
  }

  const tooltip = (
    <Primitive
      as="span"
      id={id}
      ref={popperRef}
      role="tooltip"
      style={contentStyles}
      className={cx(styles.tooltip, className)}
      testId={testId}
      onMouseEnter={() => {
        setIsHoveringContent(true);
      }}
      onMouseLeave={() => {
        setIsHoveringContent(false);
      }}
      {...attributes.popper}
    >
      <Primitive as="span">{content}</Primitive>
      <Primitive
        as="span"
        className={styles.tooltipArrow}
        data-placement={
          attributes.popper && attributes.popper['data-popper-placement']
        }
        ref={setArrowRef}
        style={popperStyles.arrow}
      />
    </Primitive>
  );

  return (
    <>
      {show ? <>{usePortal ? <Portal>{tooltip}</Portal> : tooltip}</> : null}
      <HtmlTag
        ref={elementRef}
        className={cx(styles.tooltipContainer, targetWrapperClassName)}
        onMouseEnter={(evt: MouseEvent) => {
          setIsHoveringTarget(true);
          if (onMouseOver) onMouseOver(evt);
        }}
        onMouseLeave={(evt: MouseEvent) => {
          setTimeout(() => setIsHoveringTarget(false), hideDelay);
          if (onMouseLeave) onMouseLeave(evt);
        }}
        onFocus={(evt: FocusEvent) => {
          setIsHoveringTarget(true);
          if (onFocus) onFocus(evt);
        }}
        onBlur={(evt: FocusEvent) => {
          setTimeout(() => setIsHoveringTarget(false), hideDelay);
          if (onBlur) onBlur(evt);
        }}
        onKeyDown={(evt: KeyboardEvent) => {
          if (evt.key === 'Escape') {
            setTimeout(() => setIsHoveringTarget(false), hideDelay);
          }
          if (onKeyDown) onKeyDown(evt);
        }}
        {...otherProps}
      >
        {React.Children.map<React.ReactNode, React.ReactNode>(
          children,
          (child) => {
            if (React.isValidElement(child)) {
              return React.cloneElement(child, {
                'aria-describedby': id,
              });
            }
          },
        )}
      </HtmlTag>
    </>
  );
};
