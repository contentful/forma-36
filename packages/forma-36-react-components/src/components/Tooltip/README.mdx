Tooltips are very useful to communicate extra information related to an element on the screen.

## Best practices

- Use short and clear messages as the Tooltipâ€™s content
- Even though it's possible to pass any ReactNode as a content of the Tooltip,
  we do not recommend passing very complex components try to use only [Phrasing content](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Phrasing_content)
