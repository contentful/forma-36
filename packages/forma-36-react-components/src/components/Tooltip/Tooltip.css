@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';
@import 'settings/z-index';

:root {
  --tooltip-chevron-size: calc(1rem * (10 / var(--font-base-default)));
}

.Tooltip__target-wrapper {
  display: inline-block;
  position: relative;
}

.Tooltip {
  background: var(--color-contrast-mid);
  font-family: var(--font-stack-primary);
  font-size: calc(1rem * (12 / var(--font-base-default)));
  font-weight: var(--font-weight-medium);
  font-style: normal;
  text-decoration: none;
  color: var(--color-white);
  text-align: center;
  line-height: var(--line-height-default);
  padding: calc(1rem * (8 / var(--font-base-default)))
    calc(1rem * (10 / var(--font-base-default)));
  border-radius: calc(1rem * (4 / var(--font-base-default)));
}

.Tooltip--hidden {
  visibility: hidden;
  pointer-events: none;
  z-index: var(--z-index-negative);
}

.Tooltip__arrow {
  height: var(--tooltip-chevron-size);
  width: var(--tooltip-chevron-size);
  background-color: var(--color-contrast-mid);
  z-index: var(--z-index-negative);
}
