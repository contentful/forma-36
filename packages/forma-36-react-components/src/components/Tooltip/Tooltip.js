import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import cn from 'classnames';
import InViewport from '../InViewport';
import styles from './Tooltip.css';

const TooltipContainer = ({
  children,
  setRef,
  containerElement,
  targetWrapperClassName,
  ...otherProps
}) => {
  const ContainerElement = containerElement;
  return (
    <ContainerElement
      ref={ref => setRef(ref)}
      className={cn(styles['Tooltip__target-wrapper'], targetWrapperClassName)}
      {...otherProps}
    >
      {children}
    </ContainerElement>
  );
};

TooltipContainer.propTypes = {
  children: PropTypes.array.isRequired, // Portal is an array
  setRef: PropTypes.func.isRequired,
  containerElement: PropTypes.string.isRequired,
  targetWrapperClassName: PropTypes.string,
};

TooltipContainer.defaultProps = {
  targetWrapperClassName: undefined,
};

export class Tooltip extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    targetWrapperClassName: PropTypes.string,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    id: PropTypes.string,
    onMouseLeave: PropTypes.func,
    containerElement: PropTypes.node,
    onMouseOver: PropTypes.func,
    content: PropTypes.node,
    children: PropTypes.node.isRequired,
    testId: PropTypes.string,
    place: PropTypes.oneOf(['top', 'bottom', 'right', 'left']),
    isVisible: PropTypes.bool,
    maxWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  };

  static defaultProps = {
    extraClassNames: undefined,
    targetWrapperClassName: undefined,
    onFocus: () => {},
    onBlur: () => {},
    onMouseLeave: () => {},
    onMouseOver: () => {},
    id: undefined,
    containerElement: 'span',
    isVisible: false,
    testId: 'cf-ui-tooltip',
    place: 'top',
    maxWidth: 360,
    content: undefined,
  };

  constructor(props) {
    super(props);
    this.portalTarget = document.createElement('div');
    this.place = props.place;
  }

  state = {
    isVisible: this.props.isVisible,
  };

  componentDidMount() {
    document.body.appendChild(this.portalTarget);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.content !== this.props.content) {
      this.forceUpdate();
    }
  }

  componentWillUnmount() {
    document.body.tooltipRoot.removeChild(this.portalTarget);
  }

  setPlace = place => {
    this.place = place;
  };

  calculatePosition = () => {
    if (!this.containerDomNode || !this.tooltipDomNode) {
      return { top: null, left: null };
    }

    let calculatedPosition = {};
    const containerRect = this.containerDomNode.getBoundingClientRect();
    const tooltipRect = this.tooltipDomNode.getBoundingClientRect();
    const carretVerticalOffset = 20;
    const carretHorizontalOffset = 12;

    switch (this.place) {
      case 'top':
        calculatedPosition = {
          left:
            containerRect.left +
            (containerRect.width / 2 - tooltipRect.width / 2),
          top: containerRect.top - (tooltipRect.height + carretVerticalOffset),
        };
        break;
      case 'bottom':
        calculatedPosition = {
          left:
            containerRect.left +
            (containerRect.width / 2 - tooltipRect.width / 2),
          top: containerRect.top + containerRect.height,
        };
        break;
      case 'left':
        calculatedPosition = {
          left:
            containerRect.left - (tooltipRect.width + carretHorizontalOffset),
          top: containerRect.top - tooltipRect.height / 2,
        };
        break;
      case 'right':
        calculatedPosition = {
          left:
            containerRect.left + (containerRect.width + carretHorizontalOffset),
          top: containerRect.top - tooltipRect.height / 2,
        };
        break;
      default:
        calculatedPosition = {};
    }
    return calculatedPosition;
  };

  renderTooltip = content => {
    const placeClass = `Tooltip--place-${this.place}`;
    const classNames = cn(
      styles.Tooltip,
      styles[placeClass],
      this.props.extraClassNames,
      {
        [styles['Tooltip--hidden']]: !this.state.isVisible,
      },
    );
    const tooltip = (
      <div
        role="tooltip"
        id={this.props.id}
        aria-hidden={this.state.isVisible ? 'false' : 'true'}
        style={{
          ...this.calculatePosition(),
          maxWidth: this.props.maxWidth,
        }}
        ref={ref => {
          this.tooltipDomNode = ref;
        }}
        contentEditable={false}
        onFocus={() => {
          this.setState({ isVisible: false });
        }}
        onMouseOver={() => {
          this.setState({ isVisible: false });
        }}
        className={classNames}
        data-test-id={this.props.testId}
      >
        <InViewport
          onOverflowTop={() => this.setPlace('bottom')}
          onOverflowLeft={() => this.setPlace('right')}
          onOverflowBottom={() => this.setPlace('top')}
          onOverflowRight={() => this.setPlace('left')}
        >
          {content}
        </InViewport>
      </div>
    );

    return ReactDOM.createPortal(tooltip, this.portalTarget);
  };

  render() {
    const {
      extraClassNames,
      targetWrapperClassName,
      content,
      onMouseLeave,
      onMouseOver,
      onFocus,
      containerElement,
      onBlur,
      children,
      place,
      isVisible,
      testId,
      maxWidth,
      ...otherProps
    } = this.props;

    return (
      <TooltipContainer
        containerElement={containerElement}
        onMouseOver={e => {
          this.setState({ isVisible: true });
          onMouseOver(e);
        }}
        onMouseLeave={e => {
          this.setState({ isVisible: false });
          onMouseLeave(e);
        }}
        onFocus={e => {
          this.setState({ isVisible: true });
          onFocus(e);
        }}
        onBlur={e => {
          this.setState({ isVisible: false });
          onBlur(e);
        }}
        setRef={ref => {
          this.containerDomNode = ref;
        }}
        targetWrapperClassName={targetWrapperClassName}
        aria-describedby={this.props.id}
        {...otherProps}
      >
        {children}
        {content && this.renderTooltip(content)}
      </TooltipContainer>
    );
  }
}

export default Tooltip;
