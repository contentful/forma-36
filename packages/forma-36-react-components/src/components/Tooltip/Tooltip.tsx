import React, {
  useEffect,
  useState,
  useRef,
  MouseEvent,
  FocusEvent,
  CSSProperties,
} from 'react';
import { usePopper } from 'react-popper';
import { Placement } from '@popperjs/core';
import cn from 'classnames';
import type * as CSS from 'csstype';

import styles from './Tooltip.css';
import tokens from '@contentful/forma-36-tokens';

export type TooltipPlace = Placement;

export interface TooltipProps {
  /**
   * Child nodes to be rendered in the component and that will show the tooltip when they are hovered
   */
  children: React.ReactNode;
  /**
   * Class names to be appended to the className prop of the Tooltip wrapper
   */
  className?: string;
  /**
   * HTML element used to wrap the target of the Tooltip
   */
  containerElement?: React.ElementType;
  /**
   * Content of the Tooltip
   */
  content?: React.ReactNode;
  id?: string;
  /**
   * It controls the initial visibility of the Tooltip
   */
  isVisible?: boolean;
  /**
   * It controls whether or not the tooltip stays open when hovering on the content
   */
  closeOnMouseLeave?: boolean;
  /**
   * It sets a max-width for the Tooltip
   */
  maxWidth?: number | CSS.Property.MaxWidth;
  /**
   * Function that will be called when target gets blurred
   */
  onBlur?: (evt: FocusEvent) => void;
  /**
   * Function that will be called when target gets focused
   */
  onFocus?: (evt: FocusEvent) => void;
  /**
   * Function that will be called when the user move the mouse out of the target
   */
  onMouseLeave?: (evt: MouseEvent) => void;
  /**
   * Function that will be called when the user move the mouse over of the target
   */
  onMouseOver?: (evt: MouseEvent) => void;
  /**
   * It sets the "preferred" position of the Tooltip
   */
  place?: TooltipPlace;
  /**
   * Class names to be appended to the className prop of the Tooltipâ€™s target
   */
  targetWrapperClassName?: string;
  /**
   * An ID used for testing purposes applied as a data attribute (data-test-id)
   */
  testId?: string;
}

interface ArrowPositionState {
  top: string;
  left: string;
}

export const Tooltip = ({
  children,
  className,
  containerElement: ContainerElement = 'span',
  content,
  id,
  isVisible,
  closeOnMouseLeave,
  maxWidth,
  onBlur,
  onFocus,
  onMouseLeave,
  onMouseOver,
  place,
  targetWrapperClassName,
  testId,
  ...otherProps
}: TooltipProps) => {
  const [show, setShow] = useState(isVisible);
  const [arrowPosition, setArrowPosition] = useState<ArrowPositionState>(
    getArrowPosition('bottom'),
  );

  const elementRef = useRef(null);
  const popperRef = useRef(null);
  const [arrowRef, setArrowRef] = useState<HTMLSpanElement | null>(null);
  const { styles: popperStyles, attributes, forceUpdate } = usePopper(
    elementRef.current,
    popperRef.current,
    {
      placement: place,
      modifiers: [
        { name: 'arrow', options: { element: arrowRef } },
        {
          name: 'offset',
          options: {
            offset: [0, 10],
          },
        },
      ],
    },
  );

  // necessary to update tooltip position in case the content is being updated
  useEffect(() => {
    if (forceUpdate !== null) {
      forceUpdate();
    }
  }, [content, forceUpdate]);

  useEffect(() => {
    if (attributes.popper) {
      const newPosition = getArrowPosition(
        attributes.popper['data-popper-placement'],
      );
      setArrowPosition(newPosition);
    }
  }, [attributes.popper]);

  const [isHoveringTarget, setIsHoveringTarget] = useState(false)
  const [isHoveringContent, setIsHoveringContent] = useState(false)
  useEffect(() => {
    if (closeOnMouseLeave) {
      setShow(isHoveringTarget)
    } else {
      setShow(isHoveringContent || isHoveringTarget)
    }
  }, [closeOnMouseLeave, isHoveringTarget, isHoveringContent])

  const delay = closeOnMouseLeave ? 0 : 1000;

  const arrowStyles = {
    ...popperStyles.arrow,
    ...arrowPosition,
    transform: 'rotate(45deg)',
  };

  const contentMaxWidth =
    typeof maxWidth === 'string' ? maxWidth : `${maxWidth}px`;

  const contentStyles: CSSProperties = {
    zIndex: tokens.zIndexTooltip,
    maxWidth: contentMaxWidth,
    ...popperStyles.popper,
  };

  if (!content) {
    return (
      <ContainerElement className={targetWrapperClassName}>
        {children}
      </ContainerElement>
    );
  }

  return (
    <>
      <ContainerElement
        ref={elementRef}
        className={targetWrapperClassName}
        onMouseEnter={(evt: MouseEvent) => {
          setIsHoveringTarget(true);
          if (onMouseOver) onMouseOver(evt);
        }}
        onMouseLeave={(evt: MouseEvent) => {
          setTimeout(() => setIsHoveringTarget(false), delay);
          if (onMouseLeave) onMouseLeave(evt);
      }}
        onFocus={(evt: FocusEvent) => {
          setIsHoveringTarget(true);
          if (onFocus) onFocus(evt);
        }}
        onBlur={(evt: FocusEvent) => {
          setTimeout(() => setIsHoveringTarget(false), delay);
          if (onBlur) onBlur(evt);
        }}
        {...otherProps}
      >
        {children}
      </ContainerElement>

      {show && (
        <span
          id={id}
          ref={popperRef}
          aria-hidden={show ? 'true' : 'false'}
          role="tooltip"
          style={contentStyles}
          className={cn(styles.Tooltip, className, {
            [styles['Tooltip--hidden']]: !show,
          })}
          data-test-id={testId}
          onMouseEnter={() => {
            setIsHoveringContent(true)
          }}
          onMouseLeave={() => {
            setIsHoveringContent(false)
          }}
          {...attributes.popper}
        >
          {content}
          <span
            ref={setArrowRef}
            style={arrowStyles}
            className={styles.Tooltip__arrow}
          />
        </span>
      )}
    </>
  );
};
Tooltip.defaultProps = {
  containerElement: 'span',
  isVisible: false,
  closeOnMouseLeave: true,
  maxWidth: 360,
  testId: 'cf-ui-tooltip',
  place: 'auto',
};

function getArrowPosition(popperPlacement: string) {
  const centered = 'calc(50% - 5px)';
  const oppositeToThisSide = 'calc(100% - 5px)';
  const atThisSide = '-5px';
  const atStart = '10px';
  const atEnd = 'calc(100% - 20px)';

  // the arrow is 10x10, that's why we need the -5px to correct its center
  switch (popperPlacement) {
    case 'top':
      return { top: oppositeToThisSide, left: centered }; // arrow will be V
    case 'top-start':
      return { top: oppositeToThisSide, left: atStart };
    case 'top-end':
      return { top: oppositeToThisSide, left: atEnd };
    case 'right':
      return { top: centered, left: atThisSide }; // arrow will be <
    case 'right-start':
      return { top: atStart, left: atThisSide };
    case 'right-end':
      return { top: atEnd, left: atThisSide };
    case 'left':
      return { top: centered, left: oppositeToThisSide }; // arrow will be >
    case 'left-start':
      return { top: atStart, left: oppositeToThisSide };
    case 'left-end':
      return { top: atEnd, left: oppositeToThisSide };
    case 'bottom-start':
      return { top: atThisSide, left: atStart }; // arrow will be ^
    case 'bottom-end':
      return { top: atThisSide, left: atEnd };
    default:
      return { top: atThisSide, left: centered };
  }
}

export default Tooltip;
