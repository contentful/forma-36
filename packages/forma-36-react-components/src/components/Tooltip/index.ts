export { Tooltip } from './Tooltip';
export type { TooltipPlace, TooltipProps } from './Tooltip';
