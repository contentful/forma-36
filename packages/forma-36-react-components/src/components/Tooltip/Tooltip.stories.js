import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, select } from '@storybook/addon-knobs';

import Tooltip from './Tooltip';
import TextLink from '../TextLink';

storiesOf('Components|Tooltip', module)
  .add('default', () => (
    <div>
      <Tooltip
        place={select(
          'Initial place',
          {
            Top: 'top',
            Bottom: 'bottom',
            Left: 'left',
            Right: 'right',
          },
          'top',
        )}
        id="tip1"
        targetWrapperClassName={text('Wrapper class name', 'target-wrapper')}
        content={text('Tooltip Text', 'Hi I am a Tooltip')}
      >
        <TextLink>Hover me</TextLink>
      </Tooltip>
    </div>
  ))
  .add('with HTML', () => (
    <div style={{ fontSize: 14 }}>
      Lorem Ipsum dolor sit amet &nbsp;
      <Tooltip
        place={select(
          'Initial place',
          {
            Top: 'top',
            Bottom: 'bottom',
            Left: 'left',
            Right: 'right',
          },
          'top',
        )}
        id="tip1"
        content={
          <div>
            I have some <div style={{ color: 'red' }}>HTML</div> in me
          </div>
        }
      >
        <TextLink>Hover me</TextLink>.
      </Tooltip>
      &nbsp; Lorem Ipsum dolor sit amet.
    </div>
  ));
