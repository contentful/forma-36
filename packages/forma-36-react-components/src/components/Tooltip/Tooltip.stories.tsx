import React from 'react';

import Tooltip from './Tooltip';
import Paragraph from '../Typography/Paragraph';
import TextLink from '../TextLink';
import notes from './README.md';

export default {
  title: 'Components/Tooltip',
  component: Tooltip,
  parameters: {
    propTypes: Tooltip['__docgenInfo'],
    notes,
  },
  argTypes: {
    content: { control: 'text' },
    children: { control: { disable: true } },
    className: { control: { disable: true } },
    testId: { control: { disable: true } },
  },
};

export const basic = (args: { content: string }) => {
  return (
    <Tooltip {...args}>
      <TextLink>Hover me</TextLink>
    </Tooltip>
  );
};
basic.args = {
  content: 'I am a Tooltip ðŸ™Œ',
};
const basicSourceCode = `<Tooltip content="I am a Tooltip ðŸ™Œ">
  <TextLink>Hover me</TextLink>
</Tooltip>`;
basic.parameters = {
  docs: {
    source: {
      code: basicSourceCode,
    },
  },
};

export const withHtml = (args: { content: string }) => {
  return (
    <Paragraph>
      Lorem Ipsum dolor sit amet&nbsp;
      <Tooltip {...args}>
        <TextLink>Hover me</TextLink>.
      </Tooltip>
      &nbsp; Lorem Ipsum dolor sit amet.
    </Paragraph>
  );
};
withHtml.args = {
  content: (
    <>
      I have some <b style={{ color: 'red' }}>HTML</b> in me
    </>
  ),
};
const withHtmlSourceCode = `<Tooltip content={(
  <>
    I have some <b style={{ color: 'red' }}>HTML</b> in me
  </>
)}>
  <TextLink>Hover me</TextLink>
</Tooltip>`;
withHtml.parameters = {
  docs: {
    source: {
      code: withHtmlSourceCode,
    },
  },
};

export const autoPlacement = (args: { content: string }) => {
  return (
    <div
      style={{
        height: '300px',
        width: '100%',
        overflowY: 'scroll',
      }}
    >
      <div
        style={{
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'center',
          alignItems: 'center',
          height: '600px',
          width: '100%',
        }}
      >
        <Tooltip {...args}>
          <TextLink>Hover me</TextLink>
        </Tooltip>
      </div>
    </div>
  );
};
autoPlacement.args = {
  place: 'auto',
  content: (
    <>
      I will reposition automatically
      <br />
      when you scroll
    </>
  ),
};
const autoPlacementSourceCode = `<Tooltip place="auto" content={<>I will reposition automatically<br/>when you scroll</>}>
  <TextLink>Hover me</TextLink>
</Tooltip>`;
autoPlacement.parameters = {
  docs: {
    source: {
      code: autoPlacementSourceCode,
    },
  },
};

export const contentHover = (args: { content: string }) => {
  return (
    <Paragraph>
      Lorem Ipsum dolor sit amet&nbsp;
      <Tooltip {...args}>
        <TextLink>Hover me</TextLink>.
      </Tooltip>
      &nbsp; Lorem Ipsum dolor sit amet.
    </Paragraph>
  );
};
contentHover.args = {
  closeOnMouseLeave: false,
  content: (
    <>
      You can interact with the content in me
      <br />
      <button type="button">Button</button>
      <br />
      <a style={{ color: 'white' }} href="/" target="_blank" rel="noopener noreferrer">Click me!</a> 
    </>
  ),
};
const contentHoverSourceCode = `<Tooltip closeOnMouseLeave={false} content={(
  <>
    You can interact with the content in me
    <br />
    <button type="button">Button</button>
    <br />
    <a style={{ color: 'white' }} href="/" target="_blank" rel="noopener noreferrer">Click me!</a> 
  </>
)}>
  <TextLink>Hover me</TextLink>
</Tooltip>`;
contentHover.parameters = {
  docs: {
    source: {
      code: contentHoverSourceCode,
    },
  },
};
