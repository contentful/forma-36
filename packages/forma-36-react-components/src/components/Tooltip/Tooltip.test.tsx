import React from 'react';
import { mount } from 'enzyme';
import axe from '../../utils/axeHelper';
import Tooltip from './Tooltip';

it('does not render the component if no mouseover event on child', () => {
  const output = mount(
    <Tooltip content="Tooltip content">
      <span>Hover me</span>
    </Tooltip>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component', () => {
  const output = mount(
    <Tooltip content="Tooltip content">
      <span id="test">Hover me</span>
    </Tooltip>,
  );

  output.find('#test').simulate('mouseover');
  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = mount(
    <Tooltip content="Tooltip content" className="extra-class-name">
      <span id="test">Hover me</span>
    </Tooltip>,
  );

  output.find('#test').simulate('mouseover');
  expect(output).toMatchSnapshot();
});

it('renders the component with a target wrapper classname', () => {
  const output = mount(
    <Tooltip
      content="Tooltip content"
      targetWrapperClassName="target-wrapper-class-name"
    >
      <span id="test">Hover me</span>
    </Tooltip>,
  );

  output.find('#test').simulate('mouseover');
  expect(output).toMatchSnapshot();
});

it('renders the component with a place attribute', () => {
  const output = mount(
    <Tooltip content="Tooltip content" place="left">
      <span id="test">Hover me</span>
    </Tooltip>,
  );

  output.find('#test').simulate('mouseover');
  expect(output).toMatchSnapshot();
});

it('renders the component with a id attribute', () => {
  const output = mount(
    <Tooltip id="Tooltip" content="Tooltip content">
      <span id="test">Hover me</span>
    </Tooltip>,
  );

  output.find('#test').simulate('mouseover');
  expect(output).toMatchSnapshot();
});

it('renders the component as span with a id attribute', () => {
  const output = mount(
    <Tooltip containerElement="span" content="Tooltip content">
      <span id="test">Hover me</span>
    </Tooltip>,
  );

  output.find('#test').simulate('mouseover');
  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(
    <Tooltip id="Tooltip" content="Tooltip content">
      <span>Hover me</span>
    </Tooltip>,
  ).html();

  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
