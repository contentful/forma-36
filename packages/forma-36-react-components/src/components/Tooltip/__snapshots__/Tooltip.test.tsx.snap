// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render the component if no mouseover event on child 1`] = `
<Tooltip
  containerElement="span"
  content="Tooltip content"
  isVisible={false}
  maxWidth={360}
  place="bottom"
  testId="cf-ui-tooltip"
>
  <span
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span>
      Hover me
    </span>
  </span>
</Tooltip>
`;

exports[`renders the component 1`] = `
<Tooltip
  containerElement="span"
  content="Tooltip content"
  isVisible={false}
  maxWidth={360}
  place="bottom"
  testId="cf-ui-tooltip"
>
  <span
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      id="test"
    >
      Hover me
    </span>
  </span>
</Tooltip>
`;

exports[`renders the component as span with a id attribute 1`] = `
<Tooltip
  containerElement="span"
  content="Tooltip content"
  isVisible={false}
  maxWidth={360}
  place="bottom"
  testId="cf-ui-tooltip"
>
  <span
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      id="test"
    >
      Hover me
    </span>
  </span>
</Tooltip>
`;

exports[`renders the component with a id attribute 1`] = `
<Tooltip
  containerElement="span"
  content="Tooltip content"
  id="Tooltip"
  isVisible={false}
  maxWidth={360}
  place="bottom"
  testId="cf-ui-tooltip"
>
  <span
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      id="test"
    >
      Hover me
    </span>
  </span>
</Tooltip>
`;

exports[`renders the component with a place attribute 1`] = `
<Tooltip
  containerElement="span"
  content="Tooltip content"
  isVisible={false}
  maxWidth={360}
  place="left"
  testId="cf-ui-tooltip"
>
  <span
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      id="test"
    >
      Hover me
    </span>
  </span>
</Tooltip>
`;

exports[`renders the component with a target wrapper classname 1`] = `
<Tooltip
  containerElement="span"
  content="Tooltip content"
  isVisible={false}
  maxWidth={360}
  place="bottom"
  targetWrapperClassName="target-wrapper-class-name"
  testId="cf-ui-tooltip"
>
  <span
    className="target-wrapper-class-name"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      id="test"
    >
      Hover me
    </span>
  </span>
</Tooltip>
`;

exports[`renders the component with an additional class name 1`] = `
<Tooltip
  className="extra-class-name"
  containerElement="span"
  content="Tooltip content"
  isVisible={false}
  maxWidth={360}
  place="bottom"
  testId="cf-ui-tooltip"
>
  <span
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      id="test"
    >
      Hover me
    </span>
  </span>
</Tooltip>
`;
